src/det.jl

```diff
@@ -24,18 +25,25 @@ end
     return vecdot(x0, cross(x1, x2))
 end
 
-@generated function _det{S}(::Size{S}, A::StaticMatrix)
-    if S[1] != S[2]
-        throw(DimensionMismatch("matrix is not square"))
-    end
-    return quote # Implementation from Base
-        @_inline_meta
-        T = eltype(A)
-        T2 = typeof((one(T)*zero(T) + zero(T))/one(T))
-        if istriu(A) || istril(A)
-            return convert(T2, det(UpperTriangular(A))) # Is this a Julia bug that a convert is not type stable??
+@inline _logdet(S::Union{Size{(1,1)},Size{(2,2)},Size{(3,3)}}, A::StaticMatrix) = log(_det(S, A))
+
+for (symb, f) in [(:_det, :det), (:_logdet, :logdet)]
+    eval(quote
+        @generated function $symb{S}(::Size{S}, A::StaticMatrix)
+            if S[1] != S[2]
+                throw(DimensionMismatch("matrix is not square"))
+            end
+            return quote # Implementation from Base
+                @_inline_meta
+                T = eltype(A)
+                T2 = typeof((one(T)*zero(T) + zero(T))/one(T))
+                if istriu(A) || istril(A)
+                    return convert(T2, $($f)(UpperTriangular(A))) # Is this a Julia bug that a convert is not type stable??
```

Comment:
`convert` is not treated differently to any other method, so in theory it doesn't have to return `T2`, and inference must infer that it does.